source/opt/common_uniform_elim_pass.cpp

```diff
@@ -0,0 +1,613 @@
+    }
+    for (auto ii = bp->begin(); ii != bp->end(); ++ii) {
+      if (ii->opcode() != SpvOpLoad)
+        continue;
+      uint32_t varId;
+      ir::Instruction* ptrInst = GetPtr(&*ii, &varId);
+      if (ptrInst->opcode() != SpvOpVariable)
+        continue;
+      if (!IsUniformVar(varId))
+        continue;
+      if (IsSamplerOrImageVar(varId))
+        continue;
+      if (HasUnsupportedDecorates(ii->result_id()))
+        continue;
+      uint32_t replId;
+      const auto uItr = uniform2load_id_.find(varId);
+      if (uItr != uniform2load_id_.end()) {
+        replId = uItr->second;
+      }
+      else {
+        if (mergeBlockId == 0) {
+          // Load is in dominating block; just remember it
+          uniform2load_id_[varId] = ii->result_id();
+          continue;
+        }
+        else {
+          // Copy load into most recent dominating block and remember it
+          replId = TakeNextId();
+          std::unique_ptr<ir::Instruction> newLoad(new ir::Instruction(SpvOpLoad,
+            ii->type_id(), replId, {{spv_operand_type_t::SPV_OPERAND_TYPE_ID, {varId}}}));
+          def_use_mgr_->AnalyzeInstDefUse(&*newLoad);
+          insertItr = insertItr.InsertBefore(std::move(newLoad));
+          ++insertItr;
+          uniform2load_id_[varId] = replId;
+        }
+      }
+      ReplaceAndDeleteLoad(&*ii, replId, ptrInst);
+      modified = true;
+    }
+    // If we are outside of any control construct and entering one, remember
+    // the id of the merge block
+    if (mergeBlockId == 0) {
+      uint32_t dummy;
+      mergeBlockId = MergeBlockIdIfAny(*bp, &dummy);
+    }
+  }
+  return modified;
+}
+
+bool CommonUniformElimPass::CommonUniformLoadElimBlock(ir::Function* func) {
+  bool modified = false;
+  for (auto& blk : *func) {
+    uniform2load_id_.clear();
+    for (auto ii = blk.begin(); ii != blk.end(); ++ii) {
+      if (ii->opcode() != SpvOpLoad)
+        continue;
+      uint32_t varId;
+      ir::Instruction* ptrInst = GetPtr(&*ii, &varId);
+      if (ptrInst->opcode() != SpvOpVariable)
+        continue;
+      if (!IsUniformVar(varId))
+        continue;
+      if (!IsSamplerOrImageVar(varId))
+        continue;
+      if (HasUnsupportedDecorates(ii->result_id()))
+        continue;
+      uint32_t replId;
+      const auto uItr = uniform2load_id_.find(varId);
+      if (uItr != uniform2load_id_.end()) {
+        replId = uItr->second;
+      }
+      else {
+        uniform2load_id_[varId] = ii->result_id();
+        continue;
+      }
+      ReplaceAndDeleteLoad(&*ii, replId, ptrInst);
+      modified = true;
+    }
+  }
+  return modified;
+}
+
+bool CommonUniformElimPass::CommonExtractElimination(ir::Function* func) {
+  // Find all composite ids with duplicate extracts.
+  for (auto bi = func->begin(); bi != func->end(); ++bi) {
+    for (auto ii = bi->begin(); ii != bi->end(); ++ii) {
+      if (ii->opcode() != SpvOpCompositeExtract)
+        continue;
+      if (ii->NumInOperands() > 2)
```

Comment:
Yes, this is a simplification and yes, I will add a comment. There is some opportunity for multiple indices, but it is in the noise. It would also complicate the (already complicated) data structure I am using.